File: Sources/EdgeChatCore/MemoryIndex.swift
import Foundation
import NaturalLanguage

/// A retrievable chunk of conversation history (a message, or a slice of a long message/document).
public struct MemoryChunk: Codable, Hashable, Sendable, Identifiable {
    public var id: UUID
    public var messageID: UUID
    public var messageIndex: Int
    public var role: Role
    public var source: String          // "message" or the attachment file name
    public var text: String
    public var embedding: [Float]      // empty when the embedder had nothing to say
    public var createdAt: Date

    public init(id: UUID, messageID: UUID, messageIndex: Int, role: Role, source: String, text: String, embedding: [Float], createdAt: Date) {
        self.id = id; self.messageID = messageID; self.messageIndex = messageIndex; self.role = role
        self.source = source; self.text = text; self.embedding = embedding; self.createdAt = createdAt
    }

    // Vectors are stored as little-endian Float32 base64 (~4x smaller and much faster to parse than a JSON number array).
    private enum CodingKeys: String, CodingKey { case id, messageID, messageIndex, role, source, text, embedding, vector, createdAt }

    public init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = try c.decode(UUID.self, forKey: .id)
        messageID = try c.decode(UUID.self, forKey: .messageID)
        messageIndex = try c.decode(Int.self, forKey: .messageIndex)
        role = try c.decode(Role.self, forKey: .role)
        source = try c.decode(String.self, forKey: .source)
        text = try c.decode(String.self, forKey: .text)
        createdAt = try c.decode(Date.self, forKey: .createdAt)
        if let data = try c.decodeIfPresent(Data.self, forKey: .vector) {
            embedding = data.withUnsafeBytes { raw in
                raw.bindMemory(to: UInt32.self).map { Float(bitPattern: UInt32(littleEndian: $0)) }
            }
        } else {
            embedding = try c.decodeIfPresent([Float].self, forKey: .embedding) ?? []
        }
    }

    public func encode(to encoder: Encoder) throws {
        var c = encoder.container(keyedBy: CodingKeys.self)
        try c.encode(id, forKey: .id)
        try c.encode(messageID, forKey: .messageID)
        try c.encode(messageIndex, forKey: .messageIndex)
        try c.encode(role, forKey: .role)
        try c.encode(source, forKey: .source)
        try c.encode(text, forKey: .text)
        try c.encode(createdAt, forKey: .createdAt)
        if !embedding.isEmpty {
            var data = Data(capacity: embedding.count * 4)
            for f in embedding { var v = f.bitPattern.littleEndian; withUnsafeBytes(of: &v) { data.append(contentsOf: $0) } }
            try c.encode(data, forKey: .vector)
        }
    }
}

/// Per-conversation retrieval index: on-device sentence embeddings (NaturalLanguage) + BM25, no model download.
public struct MemoryIndex: Codable, Sendable {
    public var chunks: [MemoryChunk] = []
    public var indexedMessageIDs: Set<UUID> = []
    /// Identifies the embedding space (model) the chunk vectors were produced in.
    public var embeddingKey: String?

    public init() {}

    public typealias Embedder = ([String]) async throws -> [[Float]]

    public struct Hit: Sendable {
        public let chunk: MemoryChunk
        public let score: Double
    }

    // MARK: Persistence

    public static func load(from url: URL) -> MemoryIndex {
        guard let data = try? Data(contentsOf: url), let idx = try? JSONDecoder().decode(MemoryIndex.self, from: data) else { return MemoryIndex() }
        return idx
    }

    public func save(to url: URL) throws {
        let data = try JSONEncoder().encode(self)
        try data.write(to: url, options: .atomic)
    }

    // MARK: Indexing

    /// Adds a message (and any extra texts such as full document extractions) to the index.
    /// Pass `embedder` + `embeddingKey` to compute vectors; without them chunks are lexical-only until re-embedded.
    public mutating func index(message: Message, at messageIndex: Int, extraTexts: [(source: String, text: String)] = [],
                               embedder: Embedder? = nil, embeddingKey: String? = nil) async {
        guard !indexedMessageIDs.contains(message.id) else { return }
        indexedMessageIDs.insert(message.id)
        var sources: [(String, String)] = []
        if !message.content.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty { sources.append(("message", message.content)) }
        sources += extraTexts.map { ($0.source, $0.text) }
        var new: [MemoryChunk] = []
        for (source, text) in sources {
            for piece in Self.chunk(text) {
                new.append(MemoryChunk(id: UUID(), messageID: message.id, messageIndex: messageIndex, role: message.role,
                                       source: source, text: piece, embedding: [], createdAt: message.createdAt))
            }
        }
        if let embedder, let embeddingKey, !new.isEmpty {
            if self.embeddingKey != embeddingKey { self.embeddingKey = embeddingKey; for i in chunks.indices { chunks[i].embedding = [] } }
            if let vectors = try? await embedder(new.map(\.text)), vectors.count == new.count {
                for i in new.indices { new[i].embedding = vectors[i] }
            }
        }
        chunks += new
    }

    /// Re-embeds chunks that lack a vector in the current space (bounded, so a model switch never stalls a turn).
    public mutating func refreshEmbeddings(embedder: Embedder, embeddingKey: String, limit: Int = 24, preferring ids: [UUID] = []) async {
        if self.embeddingKey != embeddingKey { self.embeddingKey = embeddingKey; for i in chunks.indices { chunks[i].embedding = [] } }
        var todo = chunks.indices.filter { chunks[$0].embedding.isEmpty }
        let preferred = Set(ids)
        todo.sort { (preferred.contains(chunks[$0].id) ? 0 : 1) < (preferred.contains(chunks[$1].id) ? 0 : 1) }
        let batch = Array(todo.prefix(limit))
        guard !batch.isEmpty, let vectors = try? await embedder(batch.map { chunks[$0].text }), vectors.count == batch.count else { return }
        for (j, i) in batch.enumerated() { chunks[i].embedding = vectors[j] }
    }

    public mutating func remove(messageID: UUID) {
        chunks.removeAll { $0.messageID == messageID }
        indexedMessageIDs.remove(messageID)
    }

    // MARK: Retrieval

    /// Hybrid search: 0.55 × cosine(model embedding) + 0.45 × normalized BM25, over `candidates`.
    /// `queryEmbedding` comes from the same embedder used for indexing (may be empty → lexical only).
    public func search(query: String, queryEmbedding qEmb: [Float], candidates: [MemoryChunk], limit: Int = 6, minScore: Double = 0.3) -> [Hit] {
        guard !candidates.isEmpty, !query.trimmingCharacters(in: .whitespaces).isEmpty else { return [] }
        let qTerms = Self.terms(query)

        // BM25 statistics over the candidate set.
        let docs = candidates.map { Self.terms($0.text) }
        let avgLen = max(1.0, Double(docs.reduce(0) { $0 + $1.count }) / Double(docs.count))
        var df: [String: Int] = [:]
        for d in docs { for t in Set(d) { df[t, default: 0] += 1 } }
        let n = Double(docs.count)
        func bm25(_ d: [String]) -> Double {
            guard !d.isEmpty else { return 0 }
            var tf: [String: Int] = [:]
            for t in d { tf[t, default: 0] += 1 }
            var score = 0.0
            for q in Set(qTerms) {
                guard let f = tf[q] else { continue }
                let dfq: Double = Double(df[q] ?? 0)
                let idf: Double = log(1 + (n - dfq + 0.5) / (dfq + 0.5))
                let tf: Double = Double(f)
                let lenNorm: Double = 0.25 + 0.75 * Double(d.count) / avgLen
                let tfn: Double = tf * 2.2 / (tf + 1.2 * lenNorm)
                score += idf * tfn
            }
            return score
        }
        let lexical = docs.map(bm25)
        let maxLex = lexical.max() ?? 0
        // Cosines from mean-pooled LLM states cluster high; rescale against the candidate spread.
        let cosines = candidates.map { (qEmb.isEmpty || $0.embedding.isEmpty) ? Double.nan : Self.cosine(qEmb, $0.embedding) }
        let valid = cosines.filter { !$0.isNaN }
        let cMin = valid.min() ?? 0, cMax = valid.max() ?? 0
        var hits: [Hit] = []
        for (i, c) in candidates.enumerated() {
            var semantic: Double = 0
            let cos: Double = cosines[i]
            if !cos.isNaN {
                if cMax > cMin {
                    semantic = (cos - cMin) / (cMax - cMin)
                } else if valid.count == 1 {
                    semantic = 1
                }
            }
            var lex: Double = 0
            if maxLex > 0 { lex = lexical[i] / maxLex }
            let score: Double = 0.55 * semantic + 0.45 * lex
            // Min-max rescaling always gives *some* chunk semantic = 1, so demand real evidence too: a shared query
            // term, or a cosine clearly at the top of a meaningful spread.
            let spread: Double = cMax - cMin
            let grounded: Bool = lexical[i] > 0 || (semantic >= 0.9 && spread >= 0.04)
            if score >= minScore, grounded { hits.append(Hit(chunk: c, score: score)) }
        }
        return Array(hits.sorted { $0.score > $1.score }.prefix(limit))
    }

    /// Renders hits as a block for the prompt, oldest first, within a character budget.
    public static func renderRecall(_ hits: [Hit], maxCharacters: Int) -> String? {
        guard !hits.isEmpty else { return nil }
        var lines: [String] = []
        var used = 0
        for h in hits.sorted(by: { $0.chunk.messageIndex < $1.chunk.messageIndex }) {
            let who = h.chunk.source == "message" ? (h.chunk.role == .user ? "User said" : "Assistant said") : "From \(h.chunk.source)"
            let line = "- (\(who), message \(h.chunk.messageIndex + 1)) \(h.chunk.text.replacingOccurrences(of: "\n", with: " "))"
            if used + line.count > maxCharacters { break }
            lines.append(line)
            used += line.count
        }
        guard !lines.isEmpty else { return nil }
        return "<recalled_context note=\"background only: earlier parts of this conversation that may be relevant. Do not repeat or re-answer them; use them only if they help with the message below.\">\n" + lines.joined(separator: "\n") + "\n</recalled_context>"
    }

    // MARK: Helpers

    /// ~600-character chunks on paragraph/sentence boundaries.
    public static func chunk(_ text: String, target: Int = 600, maxLength: Int = 900) -> [String] {
        let paragraphs = text.components(separatedBy: "\n\n").map { $0.trimmingCharacters(in: .whitespacesAndNewlines) }.filter { !$0.isEmpty }
        var out: [String] = []
        var current = ""
        func flush() { if !current.isEmpty { out.append(current); current = "" } }
        for p in paragraphs {
            let units = p.count > maxLength ? sentences(p) : [p]
            for u in units {
                if current.count + u.count + 1 > target && !current.isEmpty { flush() }
                if u.count > maxLength {
                    // Hard split very long runs (code, tables).
                    var rest = Substring(u)
                    while !rest.isEmpty {
                        let piece = String(rest.prefix(maxLength)); rest = rest.dropFirst(maxLength)
                        flush(); out.append(piece)
                    }
                } else {
                    current += current.isEmpty ? u : "\n" + u
                }
            }
        }
        flush()
        return out
    }

    private static func sentences(_ text: String) -> [String] {
        var result: [String] = []
        let tokenizer = NLTokenizer(unit: .sentence)
        tokenizer.string = text
        tokenizer.enumerateTokens(in: text.startIndex..<text.endIndex) { range, _ in
            let s = text[range].trimmingCharacters(in: .whitespacesAndNewlines)
            if !s.isEmpty { result.append(s) }
            return true
        }
        return result.isEmpty ? [text] : result
    }

    static func cosine(_ a: [Float], _ b: [Float]) -> Double {
        guard a.count == b.count, !a.isEmpty else { return 0 }
        var dot: Float = 0, na: Float = 0, nb: Float = 0
        for i in 0..<a.count { dot += a[i] * b[i]; na += a[i] * a[i]; nb += b[i] * b[i] }
        guard na > 0, nb > 0 else { return 0 }
        return Double(dot / (na.squareRoot() * nb.squareRoot()))
    }

    static func terms(_ text: String) -> [String] {
        text.lowercased()
            .components(separatedBy: CharacterSet.alphanumerics.inverted)
            .filter { $0.count > 1 && !stopwords.contains($0) }
    }

    private static let stopwords: Set<String> = ["the", "a", "an", "and", "or", "of", "to", "in", "on", "is", "it", "that", "this", "for", "with", "as", "was", "are", "be", "at", "by", "i", "you", "we", "he", "she", "they", "my", "your", "me", "do", "did", "what", "how", "why", "can", "could", "would", "should", "about", "from", "so", "if", "not", "no", "yes", "there", "here", "have", "has", "had", "will", "just", "also", "but"]
}

File: android/app/src/main/java/com/rishikesh/edgechat/data/MemoryIndex.kt
package com.rishikesh.edgechat.data

import android.util.Base64
import com.rishikesh.edgechat.model.Message
import com.rishikesh.edgechat.model.Role
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import java.io.File
import java.nio.ByteBuffer
import java.nio.ByteOrder
import java.text.BreakIterator
import java.util.UUID
import kotlin.math.ln
import kotlin.math.sqrt

/** A retrievable chunk of conversation history (a message, or a slice of a long message/document). */
@Serializable
data class MemoryChunk(
    val id: String = UUID.randomUUID().toString().uppercase(),
    val messageID: String,
    val messageIndex: Int,
    val role: Role,
    /** "message" or the attachment file name. */
    val source: String,
    val text: String,
    /** Little-endian Float32 base64 (same encoding as the iOS app); null when not embedded yet. */
    val vector: String? = null,
    val createdAt: String,
) {
    val embedding: FloatArray get() = vector?.let { decodeVector(it) } ?: FloatArray(0)
    val hasEmbedding: Boolean get() = !vector.isNullOrEmpty()

    companion object {
        fun encodeVector(v: FloatArray): String {
            val bb = ByteBuffer.allocate(v.size * 4).order(ByteOrder.LITTLE_ENDIAN)
            for (f in v) bb.putFloat(f)
            return Base64.encodeToString(bb.array(), Base64.NO_WRAP)
        }
        fun decodeVector(s: String): FloatArray {
            val bytes = runCatching { Base64.decode(s, Base64.NO_WRAP) }.getOrNull() ?: return FloatArray(0)
            val bb = ByteBuffer.wrap(bytes).order(ByteOrder.LITTLE_ENDIAN)
            return FloatArray(bytes.size / 4) { bb.getFloat() }
        }
    }
}

/** Per-conversation retrieval index: LLM mean-pooled embeddings (via the engine) + BM25. */
@Serializable
data class MemoryIndex(
    val chunks: List<MemoryChunk> = emptyList(),
    val indexedMessageIDs: Set<String> = emptySet(),
    /** Identifies the embedding space (model) the chunk vectors were produced in. */
    val embeddingKey: String? = null,
) {
    data class Hit(val chunk: MemoryChunk, val score: Double)

    /** Adds a message (and any extra texts such as full document extractions). Returns the new index. */
    suspend fun index(
        message: Message, messageIndex: Int, extraTexts: List<Pair<String, String>> = emptyList(),
        embedder: (suspend (List<String>) -> List<FloatArray>)?, embeddingKey: String?,
    ): MemoryIndex {
        if (message.id in indexedMessageIDs) return this
        val sources = mutableListOf<Pair<String, String>>()
        if (message.content.isNotBlank()) sources += "message" to message.content
        sources += extraTexts
        var new = sources.flatMap { (source, text) ->
            chunk(text).map { MemoryChunk(messageID = message.id, messageIndex = messageIndex, role = message.role, source = source, text = it, createdAt = message.createdAt) }
        }
        var chunksOut = chunks
        var key = this.embeddingKey
        if (embedder != null && embeddingKey != null && new.isNotEmpty()) {
            if (key != embeddingKey) { key = embeddingKey; chunksOut = chunksOut.map { it.copy(vector = null) } }
            val vectors = runCatching { embedder(new.map { it.text }) }.getOrNull()
            if (vectors != null && vectors.size == new.size) new = new.mapIndexed { i, c -> c.copy(vector = MemoryChunk.encodeVector(vectors[i])) }
        }
        return copy(chunks = chunksOut + new, indexedMessageIDs = indexedMessageIDs + message.id, embeddingKey = key)
    }

    /** Re-embeds chunks that lack a vector in the current space (bounded, so a model switch never stalls a turn). */
    suspend fun refreshEmbeddings(embedder: suspend (List<String>) -> List<FloatArray>, embeddingKey: String, limit: Int = 24): MemoryIndex {
        var chunksOut = chunks
        if (this.embeddingKey != embeddingKey) chunksOut = chunksOut.map { it.copy(vector = null) }
        val todo = chunksOut.indices.filter { !chunksOut[it].hasEmbedding }.take(limit)
        if (todo.isEmpty()) return copy(chunks = chunksOut, embeddingKey = embeddingKey)
        val vectors = runCatching { embedder(todo.map { chunksOut[it].text }) }.getOrNull()
        if (vectors == null || vectors.size != todo.size) return copy(chunks = chunksOut, embeddingKey = embeddingKey)
        val mutable = chunksOut.toMutableList()
        todo.forEachIndexed { j, i -> mutable[i] = mutable[i].copy(vector = MemoryChunk.encodeVector(vectors[j])) }
        return copy(chunks = mutable, embeddingKey = embeddingKey)
    }

    /** Hybrid ranking: min-max rescaled cosine (0.55) + normalized BM25 (0.45), grounded by a shared term or a clear spread winner. */
    fun search(query: String, queryEmbedding: FloatArray, candidates: List<MemoryChunk>, limit: Int = 4, minScore: Double = 0.35): List<Hit> {
        if (candidates.isEmpty() || query.isBlank()) return emptyList()
        val qTerms = terms(query)
        val docs = candidates.map { terms(it.text) }
        val avgLen = maxOf(1.0, docs.sumOf { it.size }.toDouble() / docs.size)
        val df = HashMap<String, Int>()
        for (d in docs) for (t in d.toSet()) df[t] = (df[t] ?: 0) + 1
        val n = docs.size.toDouble()
        fun bm25(d: List<String>): Double {
            if (d.isEmpty()) return 0.0
            val tf = HashMap<String, Int>()
            for (t in d) tf[t] = (tf[t] ?: 0) + 1
            var score = 0.0
            for (q in qTerms.toSet()) {
                val f = tf[q] ?: continue
                val dfq = (df[q] ?: 0).toDouble()
                val idf = ln(1 + (n - dfq + 0.5) / (dfq + 0.5))
                val lenNorm = 0.25 + 0.75 * d.size / avgLen
                score += idf * (f * 2.2 / (f + 1.2 * lenNorm))
            }
            return score
        }
        val lexical = docs.map(::bm25)
        val maxLex = lexical.maxOrNull() ?: 0.0
        val cosines = candidates.map { c -> if (queryEmbedding.isEmpty() || !c.hasEmbedding) Double.NaN else cosine(queryEmbedding, c.embedding) }
        val valid = cosines.filter { !it.isNaN() }
        val cMin = valid.minOrNull() ?: 0.0
        val cMax = valid.maxOrNull() ?: 0.0
        val hits = mutableListOf<Hit>()
        for ((i, c) in candidates.withIndex()) {
            val cos = cosines[i]
            val semantic = when {
                cos.isNaN() -> 0.0
                cMax > cMin -> (cos - cMin) / (cMax - cMin)
                valid.size == 1 -> 1.0
                else -> 0.0
            }
            val lex = if (maxLex > 0) lexical[i] / maxLex else 0.0
            val score = 0.55 * semantic + 0.45 * lex
            val grounded = lexical[i] > 0 || (semantic >= 0.9 && cMax - cMin >= 0.04)
            if (score >= minScore && grounded) hits += Hit(c, score)
        }
        return hits.sortedByDescending { it.score }.take(limit)
    }

    companion object {
        private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

        fun load(file: File): MemoryIndex =
            if (file.exists()) runCatching { json.decodeFromString<MemoryIndex>(file.readText()) }.getOrDefault(MemoryIndex()) else MemoryIndex()

        fun save(index: MemoryIndex, file: File) {
            runCatching {
                val tmp = File(file.parentFile, file.name + ".tmp")
                tmp.writeText(json.encodeToString(serializer(), index))
                if (!tmp.renameTo(file)) { file.delete(); tmp.renameTo(file) }
            }
        }

        /** Renders hits as a block for the prompt, oldest first, within a character budget. */
        fun renderRecall(hits: List<Hit>, maxCharacters: Int): String? {
            if (hits.isEmpty()) return null
            val lines = mutableListOf<String>()
            var used = 0
            for (h in hits.sortedBy { it.chunk.messageIndex }) {
                val who = if (h.chunk.source == "message") (if (h.chunk.role == Role.USER) "User said" else "Assistant said") else "From ${h.chunk.source}"
                val line = "- ($who, message ${h.chunk.messageIndex + 1}) ${h.chunk.text.replace("\n", " ")}"
                if (used + line.length > maxCharacters) break
                lines += line; used += line.length
            }
            if (lines.isEmpty()) return null
            return "<recalled_context note=\"background only: earlier parts of this conversation that may be relevant. Do not repeat or re-answer them; use them only if they help with the message below.\">\n" +
                lines.joinToString("\n") + "\n</recalled_context>"
        }

        /** ~600-character chunks on paragraph/sentence boundaries. */
        fun chunk(text: String, target: Int = 600, maxLength: Int = 900): List<String> {
            val paragraphs = text.split("\n\n").map { it.trim() }.filter { it.isNotEmpty() }
            val out = mutableListOf<String>()
            var current = StringBuilder()
            fun flush() { if (current.isNotEmpty()) { out += current.toString(); current = StringBuilder() } }
            for (p in paragraphs) {
                val units = if (p.length > maxLength) sentences(p) else listOf(p)
                for (u in units) {
                    if (current.length + u.length + 1 > target && current.isNotEmpty()) flush()
                    if (u.length > maxLength) {
                        var rest = u
                        while (rest.isNotEmpty()) { val piece = rest.take(maxLength); rest = rest.drop(maxLength); flush(); out += piece }
                    } else {
                        if (current.isEmpty()) current.append(u) else current.append("\n").append(u)
                    }
                }
            }
            flush()
            return out
        }

        private fun sentences(text: String): List<String> {
            val it = BreakIterator.getSentenceInstance()
            it.setText(text)
            val result = mutableListOf<String>()
            var start = it.first()
            var end = it.next()
            while (end != BreakIterator.DONE) {
                val s = text.substring(start, end).trim()
                if (s.isNotEmpty()) result += s
                start = end; end = it.next()
            }
            return result.ifEmpty { listOf(text) }
        }

        fun cosine(a: FloatArray, b: FloatArray): Double {
            if (a.size != b.size || a.isEmpty()) return 0.0
            var dot = 0f; var na = 0f; var nb = 0f
            for (i in a.indices) { dot += a[i] * b[i]; na += a[i] * a[i]; nb += b[i] * b[i] }
            if (na <= 0 || nb <= 0) return 0.0
            return (dot / (sqrt(na) * sqrt(nb))).toDouble()
        }

        fun terms(text: String): List<String> =
            text.lowercase().split(Regex("[^\\p{L}\\p{N}]+")).filter { it.length > 1 && it !in stopwords }

        private val stopwords = setOf("the", "a", "an", "and", "or", "of", "to", "in", "on", "is", "it", "that", "this", "for", "with", "as", "was", "are", "be", "at", "by", "i", "you", "we", "he", "she", "they", "my", "your", "me", "do", "did", "what", "how", "why", "can", "could", "would", "should", "about", "from", "so", "if", "not", "no", "yes", "there", "here", "have", "has", "had", "will", "just", "also", "but")
    }
}
